src/util/codec/mysql/decimal.rs

```diff
@@ -11,770 +11,2008 @@
+        while src_word_start_idx < stop {
+            write_word!(self, d.word_buf[src_word_start_idx] ^ mask, 4, written);
+            src_word_start_idx += 1;
+        }
+
+        if src_trailing_digits > 0 {
+            let i = DIG_2_BYTES[src_trailing_digits];
+            let lim = if src_frac_word_cnt < frac_word_cnt {
+                DIGITS_PER_WORD as usize
+            } else {
+                trailing_digits
+            };
+            while src_trailing_digits < lim && DIG_2_BYTES[src_trailing_digits] == i {
+                src_trailing_digits += 1;
+            }
+            let x = (d.word_buf[src_word_start_idx] /
+                     TEN_POW[DIGITS_PER_WORD as usize - src_trailing_digits]) ^
+                    mask;
+            write_word!(self, x, i as usize, written);
+        }
+
+        if frac_size > src_frac_size {
+            for _ in (src_frac_size..frac_size).zip(written..(int_size + frac_size) as usize) {
+                write_u8!(self, mask as u8, written);
+            }
         }
+        Ok(res)
     }
 }
 
 impl<T: Write> DecimalEncoder for T {}
 
-/// Return the first encoded decimal's length.
-///
-/// Please note that this function won't check if the decimal is
-/// encoded correctly.
-pub fn encoded_len(encoded: &[u8]) -> usize {
-    if encoded.is_empty() {
-        return 0;
-    }
-    let (is_neg, prefix_len) = match encoded[0] {
-        ZERO_SIGN => return 1,
-        NEGATIVE_SIGN => (true, 1 + number::I64_SIZE),
-        POSITIVE_SIGN => (false, 1 + number::I64_SIZE),
-        _ => return 0,
-    };
+pub trait DecimalDecoder: BytesDecoder {
+    fn decode_decimal(&mut self) -> Result<Decimal> {
+        if self.remaining() < 3 {
+            return Err(box_err!("decimal too short: {} < 3", self.remaining()));
+        }
+
+        let prec = try!(self.read_u8());
+        let frac_cnt = try!(self.read_u8());
 
-    if encoded.len() < prefix_len {
-        return encoded.len();
+        if prec < frac_cnt {
+            return Err(box_err!("invalid decimal, precision {} < frac_cnt {}",
+                                prec,
+                                frac_cnt));
+        }
+
+        let int_cnt = prec - frac_cnt;
+        let int_word_cnt = int_cnt / DIGITS_PER_WORD;
+        let leading_digits = (int_cnt - int_word_cnt * DIGITS_PER_WORD) as usize;
+        let frac_word_cnt = frac_cnt / DIGITS_PER_WORD;
+        let trailing_digits = (frac_cnt - frac_word_cnt * DIGITS_PER_WORD) as usize;
+        let mut int_word_to = int_word_cnt;
+        if leading_digits > 0 {
+            int_word_to += 1;
+        }
+        let mut frac_word_to = frac_word_cnt;
+        if trailing_digits > 0 {
+            frac_word_to += 1;
+        }
+        let mask = if self.peak_u8().unwrap() & 0x80 > 0 {
+            0
+        } else {
+            u32::MAX
+        };
+        let res = fix_word_cnt_err(int_word_to, frac_word_to, WORD_BUF_LEN);
```

Comment:
`TiDB` handles the error and keep processing, but TiKV just return the error and stop processing. Because currently once met an error, caller will just abort execution and propagate it all the way to RPC response, so just return early here is ok.